src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -0,0 +1,324 @@
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This type provides means for instrumenting compiled methods for dynamic analysis.
+    /// It can be combined with other <see cref="Instrumenter"/>s.
+    /// </summary>
+    internal partial class DynamicAnalysisInjector : CompoundInstrumenter
+    {
+        private readonly bool _doInstrumentation;
+        private readonly MethodSymbol _method;
+        private readonly MethodSymbol _createPayload;
+        private readonly ArrayBuilder<SourceSpan> _spansBuilder;
+        private readonly ArrayTypeSymbol _payloadType;
+        private readonly LocalSymbol _methodPayload;
+        private readonly TypeCompilationState _compilationState;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly DebugDocumentProvider _debugDocumentProvider;
+        private readonly bool _methodHasExplicitBlock;
+        private readonly SyntheticBoundNodeFactory _factory;
+
+        public DynamicAnalysisInjector(MethodSymbol method, BoundStatement methodBody, SyntheticBoundNodeFactory factory, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, Instrumenter previous)
+            : base (previous)
+        {
+            CSharpCompilation compilation = compilationState.Compilation;
+            _createPayload = GetCreatePayload(compilation, methodBody.Syntax, diagnostics);
+
+            // Do not instrument the instrumentation helpers if they are part of the current compilation (which occurs only during testing). GetCreatePayload will fail with an infinite recursion if it is instrumented.
+            // PROTOTYPE (https://github.com/dotnet/roslyn/issues/10266): It is not correct to always skip implict methods, because that will miss field initializers.
+            if ((object)_createPayload != null && !method.IsImplicitlyDeclared && !method.Equals(_createPayload))
+            {
+                _doInstrumentation = true;
+                _method = method;
+                _spansBuilder = ArrayBuilder<SourceSpan>.GetInstance();
+                TypeSymbol payloadElementType = factory.SpecialType(SpecialType.System_Boolean);
+                _payloadType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, payloadElementType);
+                _methodPayload = factory.SynthesizedLocal(_payloadType, kind: SynthesizedLocalKind.InstrumentationPayload, syntax: methodBody.Syntax); ;
+                _compilationState = compilationState;
+                _diagnostics = diagnostics;
+                _debugDocumentProvider = debugDocumentProvider;
+                _methodHasExplicitBlock = MethodHasExplicitBlock(method);
+                _factory = factory;
+            }
+        }
+
+        public BoundStatement AddInstrumentationPrologue(BoundBlock rewrittenMethodBody, ref ImmutableArray<SourceSpan> dynamicAnalysisSpans)
+        {
+            if (_doInstrumentation)
```

Comment:
This is odd. Can we factor this out of the instrumenter and not create the instrumenter in the first place when we are actually not instrumenting?